samples/python/onedim/diffusion_flame_continuation.py

```diff
@@ -54,43 +62,139 @@
 # Flame Continuation
 # ------------------
 
+# Maximum number of steps to take
+Nmax = 1000
+
+# Relative temperature defining control point locations, with 1 being the peak
+# temperature and 0 being the inlet temperature. Lower values tend to avoid solver
+# failures early on, while using higher values on the unstable branch tend to help
+# with finding solutions where the peak temperature is very low.
+initial_spacing = 0.6
+unstable_spacing = 0.95
+
+# Amount to adjust temperature at the control point each step [K]
+temperature_increment = 20.0
+max_increment = 100
+
+# Try to keep Tmax from changing more than this much each step [K]
+target_delta_Tmax = 20
+
+# Stop after this many successive errors
+max_error_count = 3
+error_count = 0
+
+# Stop after any failure if the strain rate has dropped to this fraction of the maximum
+strain_rate_tol = 0.10
+
 f.two_point_control_enabled = True
-spacing = 0.95
-temperature_increment = 10.0 # Kelvin
-maximum_temperature = []
-a_max = []
-for i in range(5000):
+f.max_time_step_count = 100
+Tmax = max(f.T)
+a_max = strain_rate = f.strain_rate('max')
+data = []  # integral output quantities for each step
+logger.info('Starting two-point control')
+
+for i in range(Nmax):
+    if strain_rate > 0.98 * a_max:
+        spacing = initial_spacing
+    else:
+        spacing = unstable_spacing
     control_temperature = np.min(f.T) + spacing*(np.max(f.T) - np.min(f.T))
-    print(f'Iteration {i}: Control temperature = {control_temperature} K')
+
+    # Store the flame state in case the iteration fails and we need to roll back
+    backup_state = f.to_array()
+
+    logger.debug(f'Iteration {i}: Control temperature = {control_temperature:.2f} K')
     f.set_left_control_point(control_temperature)
     f.set_right_control_point(control_temperature)
 
     # This decrement is what drives the two-point control. If failure
     # occurs, try decreasing the decrement.
     f.left_control_point_temperature -= temperature_increment
     f.right_control_point_temperature -= temperature_increment
+    f.clear_stats()
+
+    if (f.left_control_point_temperature < f.fuel_inlet.T + 100
+        or f.right_control_point_temperature < f.oxidizer_inlet.T + 100
+    ):
+        logger.info("Stopping because control point temperature is sufficiently close "
```

Comment:
Is this good or bad? (Presumably the former?) … logging output could be clearer.

Reply:
I'm not sure what else to say here. I thought "sufficiently close" had a positive connotation (that is, it's not "too close").

Reply:
Maybe something about tolerances being satisfied?

Reply:
The output of the default run is:
```
INFO:__main__:Traversed unstable branch down to 4.1% of the maximum strain rate.
INFO:__main__:Stopped after 134 iterations
```
It would be nice to see `logging.info` indicating `Success` or `Failure`?

Reply:
Yeah, that works.